components/net/mime_classifier.rs

```diff
@@ -33,36 +35,112 @@ pub enum NoSniffFlag {
                         }
                     }
                 }
-            }
+            },
+            LoadContext::Image => {
+                // 8.2 Sniffing an image context
+                // 1. If the supplied MIME type is an XML type, the computed MIME type is the supplied MIME type. Abort these steps.
+                // 2. Let image-type-matched be the result of executing the image type pattern matching algorithm with the resource header as the byte sequence to be matched.
+                // 3. If image-type-matched is not undefined, the computed MIME type is image-type-matched. Abort these steps.
+                // 4. The computed MIME type is the supplied MIME type.
+                match MIMEClassifier::maybe_get_media_type(supplied_type) {
+                    Some(MediaType::Xml) | Some(MediaType::Html) => supplied_type.clone(),
+                    _ => self.image_classifier.classify(data).
+                        or(supplied_type.clone())
+                }
+            },
+            LoadContext::AudioVideo => {
+                // 8.3 Sniffing an image context
+                // 1. If the supplied MIME type is an XML type, the computed MIME type is the supplied MIME type. Abort these steps.
+                // 2.  Let audio-or-video-type-matched be the result of executing the audio or video type pattern matching algorithm with the resource header as the byte sequence to be matched.
+                // 3. If audio-or-video-type-matched is not undefined, the computed MIME type is audio-or-video-type-matched. Abort these steps.
+                // 4. The computed MIME type is the supplied MIME type.
+                match MIMEClassifier::maybe_get_media_type(supplied_type) {
+                    Some(MediaType::Xml) | Some(MediaType::Html) => supplied_type.clone(),
+                    _ => self.audio_video_classifier.classify(data).
+                        or(supplied_type.clone())
+                }
+            },
+            LoadContext::Plugin => {
+                // 8.4 Sniffing in a plugin context
+                // 1. If the supplied MIME type is undefined, the computed MIME type is "application/octet-stream". [not finalized in the specs at the time of this implementation]
+                // 2. The computed MIME type is the supplied MIME type.
+                match *supplied_type {
+                    None => Some(("application".to_owned(), "octet-stream".to_owned())),
+                    _ => supplied_type.clone()
+                }
+            },
+            LoadContext::Style => {
+                // 8.5 Sniffing in a style context
+                // 1. If the supplied MIME type is undefined, [unspecified at the time of this implementation, so fallback to Browsing]
+                // 2. The computed MIME type is the supplied MIME type.
+                match *supplied_type {
+                    None => self.classify(LoadContext::Browsing, no_sniff_flag, apache_bug_flag, supplied_type, data),
```

Comment:
The actual algorithm should be something like this:

1. If the supplied MIME type is undefined, the computed MIME type is "text/css". 
2. The computed MIME type is the supplied MIME type.